Google.Api.Gax.Grpc/ClientBuilder.cs

```diff
@@ -0,0 +1,209 @@
+        /// </summary>
+        protected ClientBuilder()
+        {
+        }
+
+        /// <summary>
+        /// Validates that the builder is in a consistent state for building. For example, it's invalid to call
+        /// <see cref="Build"/> on an instance which has both JSON credentials and a credentials path specified.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The builder is in an invalid state.</exception>
+        protected virtual void Validate()
+        {
+            // If there's a call invoker, we shouldn't have any credentials-related information or an endpoint.
+            ValidateOptionExcludesOthers("CallInvoker cannot be specified with credentials settings or an endpoint", CallInvoker,
+                ChannelCredentials, CredentialsPath, JsonCredentials, Scopes, Endpoint, TokenAccessMethod);
+
+            ValidateAtMostOneNotNull("Only one source of credentials can be specified",
+                ChannelCredentials, CredentialsPath, JsonCredentials, TokenAccessMethod);
+
+            ValidateOptionExcludesOthers("Scopes are not relevant when a token access method or channel credentials are supplied", Scopes,
+                TokenAccessMethod, ChannelCredentials);
+        }
+
+        /// <summary>
+        /// Validates that at most one of the given values is not null.
+        /// </summary>
+        /// <param name="message">The message if the condition is violated.</param>
+        /// <param name="values">The values to check for nullity.</param>
+        /// <exception cref="InvalidOperationException">More than one value is null.</exception>
+        protected void ValidateAtMostOneNotNull(string message, params object[] values)
+        {
+            int notNull = values.Count(v => v == null);
+            GaxPreconditions.CheckState(notNull < 2, message);
+        }
+
+        /// <summary>
+        /// Validates that if <paramref name="controlling"/> is not null, then every value in <paramref name="values"/> is null.
+        /// </summary>
+        /// <param name="message">The message if the condition is violated.</param>
+        /// <param name="controlling">The value controlling whether or not any other value can be non-null.</param>
+        /// <param name="values">The values checked for non-nullity if <paramref name="controlling"/> is non-null.</param>
+        protected void ValidateOptionExcludesOthers(string message, object controlling, params object[] values)
+        {
+            GaxPreconditions.CheckState(controlling == null || !values.Contains(null), message);
+        }
+
+        /// <summary>
+        /// Creates a call invoker synchronously. Override this method in a concrete builder type if more
+        /// call invoker mechanisms are supported.
+        /// This implementation calls <see cref="GetChannelCredentials"/> if no call invoker is specified.
+        /// </summary>
+        protected virtual CallInvoker CreateCallInvoker()
```

Comment:
Seems a shame this `CreateCallInvoker` and `GetChannelCredentials` sync methods are required. I guess they probably are though.

Reply:
Yup, but at least it's only effectively internal.